src/approaches/grammar_search_invention_approach.py

```diff
@@ -49,44 +50,109 @@ def _generate(self) -> Iterator[Tuple[Predicate, float]]:
+class _SingleAttributeCompareClassifier:
+    """Compare a single feature value with a constant value.
     """
     object_index: int
     attribute_name: str
-    value: float
+    constant: float
+    compare: Callable[[float, float], bool]
+    compare_str: str
 
     def __call__(self, s: State, o: Sequence[Object]) -> bool:
         obj = o[self.object_index]
-        return s.get(obj, self.attribute_name) >= self.value
+        return self.compare(s.get(obj, self.attribute_name), self.constant)
+
+    def __str__(self) -> str:
+        return f"({self.object_index}.{self.attribute_name}" + \
+               f"{self.compare_str}{self.constant:.3})"
 
 
 @dataclass(frozen=True, eq=False, repr=False)
 class _HoldingDummyPredicateGrammar(_PredicateGrammar):
     """A hardcoded cover-specific grammar.
+
     Good for testing with:
         python src/main.py --env cover --approach grammar_search_invention \
             --seed 0 --excluded_predicates Holding
     """
     def _generate(self) -> Iterator[Tuple[Predicate, float]]:
         # A necessary predicate
-        name = "InventedHolding"
         block_type = [t for t in self.types if t.name == "block"][0]
         types = [block_type]
-        classifier = _SingleAttributeGEClassifier(0, "grasp", -0.9)
-        yield (Predicate(name, types, classifier), 1.)
+        classifier = _SingleAttributeCompareClassifier(0, "grasp", -0.9,
+                                                       ge, ">=")
+        yield (Predicate(str(classifier), types, classifier), 1.)
 
         # An unnecessary predicate (because it's redundant)
-        name = "InventedDummy"
-        block_type = [t for t in self.types if t.name == "block"][0]
-        types = [block_type]
-        classifier = _SingleAttributeGEClassifier(0, "is_block", 0.5)
-        yield (Predicate(name, types, classifier), 1.)
+        classifier = _SingleAttributeCompareClassifier(0, "is_block", 0.5,
+                                                       ge, ">=")
+        yield (Predicate(str(classifier), types, classifier), 1.)
+
+
+def _halving_constant_generator(lo: float, hi: float) -> Iterator[float]:
+    mid = (hi + lo) / 2.
+    yield mid
+    left_gen = _halving_constant_generator(lo, mid)
+    right_gen = _halving_constant_generator(mid, hi)
+    for l, r in zip(left_gen, right_gen):
+        yield l
+        yield r
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _SingleFeatureInequalitiesPredicateGrammar(_PredicateGrammar):
+    """Generates features of the form "0.feature >= c" or "0.feature <= c".
+    """
+    def _generate(self) -> Iterator[Tuple[Predicate, float]]:
+        # Get ranges of feature values from data.
+        feature_ranges = self._get_feature_ranges()
+        # 0., 1., 0.5, 0.25, 0.75, 0.125, 0.375, ...
+        constant_generator = itertools.chain([0., 1.],
+            _halving_constant_generator(0., 1.))
+        for c in constant_generator:
```

Comment:
I think a better way to do this would be a KDTree?

Reply:
or something.... /shrug. probably fine as is for now.

Reply:
yeah doesn't matter, it's not even clear we'd be using it later